homework-submission/week2/src/App.js

```diff
@@ -0,0 +1,59 @@
+import React, { Component } from 'react';
+import './App.css';
+import ListHeader from './components/ListHeader';
+import List from './components/List';
+import TodoItems from './sources/todoItems.json';
+
+class App extends Component {
+  state = {
+    items: localStorage.items !== undefined ? JSON.parse(localStorage.items) : TodoItems,
+    newItem: {
+      newId: localStorage.items !== undefined ? (JSON.parse(localStorage.items).length + 1) : (TodoItems.length + 1),
+      newDescription: '',
+      newDeadLine: '',
+      newState: false,
+    }
+  };
+
+  handleChecked = (id) => {
+    let newState = Object.assign({}, this.state);
+    const indexOfItem = newState.items.findIndex(item => item.id === id);
+    let targetItem = newState.items[indexOfItem];
+    if (indexOfItem >= 0) {
+      targetItem.done === true
+        ? (targetItem.done = false)
+        : (targetItem.done = true);
+    }
+    localStorage.setItem("items", JSON.stringify(newState.items));
+    this.setState({ items: newState.items });
```

Comment:
Hi Jawhar,
Thank you very much for your feedback. It is a very useful tip.

Reply:
Sorry being late, I didn't saw this reply till the day.
yes you are mutating the state; because you have in line 19 a, `Object.assign` which in your case is just assigning a new reference to the old one.

according to [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)
> The Object.assign() method is used to copy the **values** of **all enumerable own properties** from one or more source objects to a target object. It will return the target object.

in this case it's only copying the values of the Object as a references 💔.
what you may do for a deeper copy is by initializing a new source Object in the third parameter so it will be a complete new Object:
```javascript
const setState = Object.assign({}, this.state, {/* the object with the new values */})
```

Js is wired sometimes I have no 100% correct explanation for what is happening here more than What I said.